packages/common-gax/src/service.js

```diff
@@ -0,0 +1,172 @@
+/*!
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*!
+ * @module common-gax/service
+ */
+
+'use strict';
+
+var common = require('@google-cloud/common');
+var extend = require('extend');
+var streamEvents = require('stream-events');
+var through = require('through2');
+var util = require('util');
+
+/**
+ * GaxService is a base class, meant to be inherited from by a "service" that
+ * uses GAX.
+ *
+ * @constructor
+ * @alias module:common-gax/service
+ *
+ * @param {object} config - Configuration object.
+ * @param {*} config.module - The generated module.
+ * @param {object} options - [Configuration object](#/docs/?method=gcloud).
+ */
+function GaxService(config, options) {
+  if (global.GCLOUD_SANDBOX_ENV) {
+    return;
+  }
+
+  config = extend({
+    scopes: config.module.ALL_SCOPES
+  }, config);
+
+  this.api = {};
+  this.module = config.module;
+  this.options = options;
+
+  common.Service.call(this, config, options);
+}
+
+util.inherits(GaxService, common.Service);
+
+/**
+ * Funnel all API requests through this method, to be sure we have a project ID.
+ *
+ * @param {object} config - Configuration object.
+ * @param {string} config.client - The gax client name.
+ * @param {object} config.gaxOpts - GAX options.
+ * @param {function} config.method - The gax method to call.
+ * @param {object} config.reqOpts - Request options.
+ * @param {function=} callback - The callback function.
+ */
+GaxService.prototype.request = function(config, callback) {
+  if (global.GCLOUD_SANDBOX_ENV) {
+    return;
+  }
+
+  this.prepareRequest_(config, function(err, requestFn) {
```

Comment:
It's because requestStream also needs a "prepared" request. To define prepared:

- Instantiates & caches GAX client
- Auto-inserts the project ID to the request options